flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergSink.java

```diff
@@ -0,0 +1,672 @@
+ * writers that can write data to files in parallel and route commit info globally to one Committer.
+ * Post commit topology will take of compacting the already written files and updating the file log
+ * after the compaction.
+ *
+ * <pre>{@code
+ *                            Flink sink
+ *               +------------------------------------------------------------------+
+ *               |                                                                  |
+ * +-------+     | +-------------+       +---------------+                          |
+ * | Map 1 | ==> | | writer 1 | =|       | committer 1   |                          |
+ * +-------+     | +-------------+       +---------------+                          |
+ *               |                 \                                                |
+ *               |                  \                                               |
+ *               |                   \                                              |
+ * +-------+     | +-------------+    \  +---------------+        +---------------+ |
+ * | Map 2 | ==> | | writer 2 | =| ---  >| committer 2   |  --->  | post commit   | |
+ * +-------+     | +-------------+       +---------------+        +---------------+ |
+ *               |                                                                  |
+ *               +------------------------------------------------------------------+
+ * }</pre>
+ */
+public class IcebergSink
+    implements StatefulSink<RowData, StreamWriterState>,
+        WithPreWriteTopology<RowData>,
+        WithPreCommitTopology<RowData, FilesCommittable>,
+        WithPostCommitTopology<RowData, FilesCommittable> {
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergSink.class);
+
+  private final TableLoader tableLoader;
+  private final Table table;
+  private final boolean overwrite;
+  private final boolean upsertMode;
+  private final DistributionMode distributionMode;
+  private final List<String> equalityFieldColumns;
+  private final List<Integer> equalityFieldIds;
+  private final FileFormat dataFileFormat;
+  private final int workerPoolSize;
+  private final long targetDataFileSize;
+  private final String uidPrefix;
+  private final Map<String, String> snapshotProperties;
+  private final RowType flinkRowType;
+  private Committer<FilesCommittable> committer;
+
+  public IcebergSink(
+      TableLoader tableLoader,
+      @Nullable Table newTable,
+      TableSchema tableSchema,
+      List<String> equalityFieldColumns,
+      @Nullable String uidPrefix,
+      ReadableConfig readableConfig,
+      Map<String, String> writeOptions,
+      Map<String, String> snapshotProperties) {
+    this.tableLoader = tableLoader;
+    this.equalityFieldColumns = equalityFieldColumns;
+    this.uidPrefix = uidPrefix == null ? "" : uidPrefix;
+    this.snapshotProperties = snapshotProperties;
+
+    Preconditions.checkNotNull(tableLoader, "Table loader shouldn't be null");
```

Comment:
nit: can we move this to the first line?

Reply:
good